canary/webrtc-c/src/media-server-storage/Common.cpp

```diff
@@ -0,0 +1,1864 @@
+                pSampleConfiguration->sampleStreamingSessionList[i]->rtcMetricsHistory.prevNumberOfBytesSent =
+                    pSampleConfiguration->rtcIceCandidatePairMetrics.rtcStatsObject.iceCandidatePairStats.bytesSent;
+                pSampleConfiguration->sampleStreamingSessionList[i]->rtcMetricsHistory.prevNumberOfBytesReceived =
+                    pSampleConfiguration->rtcIceCandidatePairMetrics.rtcStatsObject.iceCandidatePairStats.bytesReceived;
+                pSampleConfiguration->sampleStreamingSessionList[i]->rtcMetricsHistory.prevPacketsDiscardedOnSend =
+                    pSampleConfiguration->rtcIceCandidatePairMetrics.rtcStatsObject.iceCandidatePairStats.packetsDiscardedOnSend;
+            }
+        }
+    }
+
+CleanUp:
+
+    if (locked) {
+        MUTEX_UNLOCK(pSampleConfiguration->sampleConfigurationObjLock);
+    }
+
+    return retStatus;
+}
+
+STATUS pregenerateCertTimerCallback(UINT32 timerId, UINT64 currentTime, UINT64 customData)
+{
+    UNUSED_PARAM(timerId);
+    UNUSED_PARAM(currentTime);
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleConfiguration pSampleConfiguration = (PSampleConfiguration) customData;
+    BOOL locked = FALSE;
+    UINT32 certCount;
+    PRtcCertificate pRtcCertificate = NULL;
+
+    CHK_WARN(pSampleConfiguration != NULL, STATUS_NULL_ARG, "[KVS Master] pregenerateCertTimerCallback(): Passed argument is NULL");
+
+    // Use MUTEX_TRYLOCK to avoid possible dead lock when canceling timerQueue
+    if (!MUTEX_TRYLOCK(pSampleConfiguration->sampleConfigurationObjLock)) {
+        return retStatus;
+    } else {
+        locked = TRUE;
+    }
+
+    // Quick check if there is anything that needs to be done.
+    CHK_STATUS(stackQueueGetCount(pSampleConfiguration->pregeneratedCertificates, &certCount));
+    CHK(certCount != MAX_RTCCONFIGURATION_CERTIFICATES, retStatus);
+
+    // Generate the certificate with the keypair
+    CHK_STATUS(createRtcCertificate(&pRtcCertificate));
+
+    // Add to the stack queue
+    CHK_STATUS(stackQueueEnqueue(pSampleConfiguration->pregeneratedCertificates, (UINT64) pRtcCertificate));
+
+    DLOGV("New certificate has been pre-generated and added to the queue");
+
+    // Reset it so it won't be freed on exit
+    pRtcCertificate = NULL;
+
+    MUTEX_UNLOCK(pSampleConfiguration->sampleConfigurationObjLock);
+    locked = FALSE;
+
+CleanUp:
+
+    if (pRtcCertificate != NULL) {
+        freeRtcCertificate(pRtcCertificate);
+    }
+
+    if (locked) {
+        MUTEX_UNLOCK(pSampleConfiguration->sampleConfigurationObjLock);
+    }
+
+    return retStatus;
+}
+
+STATUS freeSampleConfiguration(PSampleConfiguration* ppSampleConfiguration)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleConfiguration pSampleConfiguration;
+    UINT32 i;
+    UINT64 data;
+    StackQueueIterator iterator;
+    BOOL locked = FALSE;
+
+    std::string filePath = "../toConsumer.txt";
```

Comment:
This file path should be set up as an env. We should use envs in places where both webrtc and consumer depend on the value to be the same so that we do not need to update the C files and Java files in the future which could cause sync issues in case of a miss in modifying required code.

Reply:
Do you prefer having two env vars for the whole relative paths (one for consumer, one for webrtc) or just have the env var handle the file name and leave the relative paths hardcoded?

Reply:
For now, I added a definition to the WebRTC end:
`#define STORAGE_CANARY_FIRST_FRAME_TS_PATH      "../firstFrameSentTimeStamp.txt"
`